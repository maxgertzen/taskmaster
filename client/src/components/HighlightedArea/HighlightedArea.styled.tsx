import styled from '@emotion/styled';

export const StyledHighlightedArea = styled('div')<{ isActive?: boolean }>(
  ({ theme, isActive = false }) => ({
    position: 'relative',
    padding: theme.spacing(1),
    background: 'transparent',
    borderRadius: '50%',
    width: 'fit-content',
    ...(isActive ? { zIndex: 10000 } : { zIndex: 'unset' }),
    pointerEvents: isActive ? 'all' : 'auto',
  })
);
