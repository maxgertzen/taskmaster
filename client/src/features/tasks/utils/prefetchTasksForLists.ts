import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
