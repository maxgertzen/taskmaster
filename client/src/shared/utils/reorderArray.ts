export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
