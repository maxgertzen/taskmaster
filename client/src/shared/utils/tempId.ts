export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
