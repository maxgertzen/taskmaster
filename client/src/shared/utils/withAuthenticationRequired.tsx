import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
