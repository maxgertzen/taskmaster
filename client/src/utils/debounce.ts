export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
