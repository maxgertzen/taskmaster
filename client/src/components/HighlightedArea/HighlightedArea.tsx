import { FC, ReactNode } from 'react';

import { useSpotlightStore } from '../../store/store';

import { StyledHighlightedArea } from './HighlightedArea.styled';

interface HighlightedAreaProps {
  id: string;
  children: ReactNode;
}

export const HighlightedArea: FC<HighlightedAreaProps> = ({ id, children }) => {
  const isActive = useSpotlightStore((state) => state.target) === id;

  return (
    <StyledHighlightedArea isActive={isActive} data-spotlight={id}>
      {children}
    </StyledHighlightedArea>
  );
};
