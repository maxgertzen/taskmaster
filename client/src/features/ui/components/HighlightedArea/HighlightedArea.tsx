import { FC, ReactNode } from 'react';

import { useSpotlightStore } from '@/shared/store/spotlightStore';

import { StyledHighlightedArea } from './HighlightedArea.styled';

interface HighlightedAreaProps {
  id: string;
  children: ReactNode;
}

export const HighlightedArea: FC<HighlightedAreaProps> = ({ id, children }) => {
  const isActive = useSpotlightStore((state) => state.target) === id;

  return (
    <StyledHighlightedArea isActive={isActive} data-spotlight={id}>
      {children}
    </StyledHighlightedArea>
  );
};
