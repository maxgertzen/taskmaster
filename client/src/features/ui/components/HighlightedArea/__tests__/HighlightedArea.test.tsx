import { describe, it, expect, vi, beforeEach } from 'vitest';

import { mockSpotlightStore } from '@/tests/__mocks__/storeMocks';
import { render, screen } from '@/tests/utils';

import { HighlightedArea } from '../HighlightedArea';

const mockUseSpotlightStore = vi.fn();
vi.mock('@/shared/store/spotlightStore', () => ({
  useSpotlightStore: mockSpotlightStore,
}));

describe('HighlightedArea', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([
    ['inactive', null],
    ['active', 'test-area'],
  ])('when %s', (_, spotlightTarget) => {
    it.each([
      ['light', undefined],
      ['dark', 'dark'],
    ])('renders correctly in %s theme', (_, theme) => {
      mockUseSpotlightStore.mockImplementation(() => spotlightTarget);

      const { container } = render(
        <HighlightedArea id='test-area'>
          <div>Test Content</div>
        </HighlightedArea>,
        { theme: theme as 'light' | 'dark' }
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Content Rendering', () => {
    it('renders children correctly', () => {
      mockUseSpotlightStore.mockImplementation(() => null);

      render(
        <HighlightedArea id='test-area'>
          <div data-testid='child'>Test Content</div>
        </HighlightedArea>
      );

      expect(screen.getByTestId('child')).toBeInTheDocument();
      expect(screen.getByText('Test Content')).toBeInTheDocument();
    });
  });

  describe('Spotlight Behavior', () => {
    it('sets data-spotlight attribute correctly', () => {
      mockUseSpotlightStore.mockImplementation(() => null);

      const { container } = render(
        <HighlightedArea id='test-area'>
          <div>Test Content</div>
        </HighlightedArea>
      );

      expect(container.firstChild).toHaveAttribute(
        'data-spotlight',
        'test-area'
      );
    });

    it('reflects active state when spotlight matches', () => {
      mockUseSpotlightStore.mockImplementation(() => 'test-area');

      const { container } = render(
        <HighlightedArea id='test-area'>
          <div>Test Content</div>
        </HighlightedArea>
      );

      expect(container.firstChild).toHaveAttribute(
        'data-spotlight',
        'test-area'
      );
    });

    it('reflects inactive state when spotlight differs', () => {
      mockUseSpotlightStore.mockImplementation(() => 'other-area');

      const { container } = render(
        <HighlightedArea id='test-area'>
          <div>Test Content</div>
        </HighlightedArea>
      );

      expect(container.firstChild).toHaveAttribute(
        'data-spotlight',
        'test-area'
      );
    });
  });
});
