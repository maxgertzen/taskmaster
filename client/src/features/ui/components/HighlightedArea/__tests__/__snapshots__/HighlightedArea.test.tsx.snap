// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HighlightedArea > when active > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1rqeb1j"
    data-spotlight="test-area"
  >
    <div>
      Test Content
    </div>
  </div>
</div>
`;

exports[`HighlightedArea > when active > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1rqeb1j"
    data-spotlight="test-area"
  >
    <div>
      Test Content
    </div>
  </div>
</div>
`;

exports[`HighlightedArea > when inactive > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1rqeb1j"
    data-spotlight="test-area"
  >
    <div>
      Test Content
    </div>
  </div>
</div>
`;

exports[`HighlightedArea > when inactive > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1rqeb1j"
    data-spotlight="test-area"
  >
    <div>
      Test Content
    </div>
  </div>
</div>
`;
